### ISSUE TITLE
[python] Weights is ignored when customized objective function is used

### ISSUE THREAD

**Issue by shiyu1994** [microsoft/LightGBM#5027]
## Description
Weights are not multiplied to the gradients and hessians from customized objective function in Python API.

## Reproducible example
```python
import numpy as np
import lightgbm as lgb

def fobj(preds, train_data):
    labels = train_data.get_label()
    return preds - labels, np.ones_like(labels)

def test():
    np.random.seed(123)
    num_data = 10000
    num_feature = 100
    train_X = np.random.randn(num_data, num_feature)
    train_y = np.mean(train_X, axis=-1)
    valid_X = np.random.randn(num_data, num_feature)
    valid_y = np.mean(valid_X, axis=-1)
    weights = np.random.rand(num_data)
    train_data = lgb.Dataset(train_X, train_y, weight=weights) # comment out weights will get the same output
    valid_data = lgb.Dataset(valid_X, valid_y)
    params = {
        "verbose": 2,
        "metric": "rmse",
        "learning_rate": 0.2,
        "num_trees": 20,
    }
    booster = lgb.train(train_set=train_data, valid_sets=[valid_data], valid_names=["valid"], params=params, fobj=fobj)

if __name__ == "__main__":
    test()
```

LightGBM version or commit hash:
Version 3.3.2

Command(s) you used to install LightGBM
Install from source

**Comment by jmoralez** [comment]
Moving the conversation from https://github.com/microsoft/LightGBM/pull/4925#discussion_r813010118 here. This may be intentional because the weights are available in the custom objective function through the training API and not through scikit-learn's but it'd be nice to clarify this.

**Comment by StrikerRUS** [comment]
> This may be intentional because the weights are available in the custom objective function through the training API and not through scikit-learn's but it'd be nice to clarify this.

I think this is a oversight because one form of custom **evaluation** function accepts weights in scikit-learn API:
https://github.com/microsoft/LightGBM/blob/97c8d94542b451814fdbf722eeabd9fff22796a4/python-package/lightgbm/sklearn.py#L138-L141
I guess the same can be done for custom objective function.

**Comment by jmoralez** [comment]
Hmm now I'm more confused because for the objective function weights aren't allowed https://github.com/microsoft/LightGBM/blob/97c8d94542b451814fdbf722eeabd9fff22796a4/python-package/lightgbm/sklearn.py#L105-L112
but for eval they are https://github.com/microsoft/LightGBM/blob/97c8d94542b451814fdbf722eeabd9fff22796a4/python-package/lightgbm/sklearn.py#L192-L201

**Comment by StrikerRUS** [comment]
> Hmm now I'm more confused because for the objective function weights aren't allowed

Yeah, exactly! I'm proposing to allow passing weights for the objective function.

**Comment by jmoralez** [comment]
I think it may be more user friendly to weigh things automatically. I think specifying sample weights either through the `Dataset` or a method in the sklearn API kind of implies that I want to use them to weigh my samples everywhere (grad, hess, metrics). It would be awkward that the grad and hess are weighted automatically but the metric isn't, and that if I switch to the training API I have to weigh everything myself. WDYT?

**Comment by StrikerRUS** [comment]
>  kind of implies that I want to use them to weigh my samples everywhere

Highly likely. But what if no?.. There will be no way to unweight them then. I think it's better to not weight automatically anything but allow user to choose weight or not weight.

**Comment by jmoralez** [comment]
> I think it's better to not weight automatically anything but allow user to choose weight or not weight.

I agree with you, it gives the user full control and could enable use cases like #4995, which could be achieved by weighing only the metric but not the grad and hess. So we should remove this then, right?https://github.com/microsoft/LightGBM/blob/97c8d94542b451814fdbf722eeabd9fff22796a4/python-package/lightgbm/sklearn.py#L115-L122

**Comment by shiyu1994** [comment]
> Highly likely. But what if no?.. There will be no way to unweight them then. I think it's better to not weight automatically anything but allow user to choose weight or not weight.

Does that mean, in the example above, we should let the user to weight the gradients in the `fobj` function?

**Comment by StrikerRUS** [comment]
> So we should remove this then, right?

> Does that mean, in the example above, we should let the user to weight the gradients in the `fobj` function?

I guess so.

**Comment by shiyu1994** [comment]
Shall we open an PR to remove the weighting in sklearn API?

**Comment by StrikerRUS** [comment]
I'm for it.
